Stubbed out CRUD endpoints for "projects" collection using express and mongoose

rexposadas opened the pull request:
# What does this PR do? 

Created simple, but working, api endpoints for creating, getting and updating "projects". 



# List steps on how to test your PR.

1 - start up mongodb with docker as documented in the wiki, but expose the port with -p 27017:27017
2 - export the db url : 
         export DATABASE_URL="mongodb://localhost:27017/test" 
3 - at the top level of this repo run `npm start` 

make some calls to the server:
create a project

```bash
curl --location 'http://localhost:3000/projects' \
--header 'Content-Type: application/json' \
--data '{
    "name": "my project"
}'
```

get projects

```bash
curl --location 'http://localhost:3000/projects'
```

update projects:
```bash
curl --location --request PUT 'http://localhost:3000/projects/63ff86a02b7c1df36c620ac4' \
--header 'Content-Type: application/json' \
--data '{
    "name":"from body update"
}'

```


# Anything else the reviewer should know about?

wbraynen commented on server_old.js:
```diff
@@ -0,0 +1,88 @@
+
+
+// A function to accept a CSV from a REST endpoint and parse it.
+function uploadFunctions(req, res) {
```
Would a better name be `uploadCsvFile`?

rexposadas replied:
I meant to delete this file. It's deprecated.

wbraynen commented on server_old.js:
```diff
@@ -0,0 +1,88 @@
+
+// A function to accept a CSV from a REST endpoint and parse it.
+function uploadFunctions(req, res) {
+    const csv = require('csv-parser')
+    var results = []
+
+
+  fs.createReadStream(req.file.path)
+        .pipe(csv())
+        .on('data', (data) => results.push(data))
+        .on('end', () => {
+          db.collection('Electrons').insert(results, function (err, result) {
+            if (err) {
+              res.status(500).json({ err })
+            } else {
+              res.status(201).json({ result })
+            }
+          })
+        });
```
Nitpick: lint?

(Indentation seems off and, also, line 41 can can be removed.)

wbraynen commented on server_old.js:
```diff
@@ -0,0 +1,88 @@
+  const spin = 1;
+  const min = 20;
+  const max = 120;
+  const inc = 20;
+  db.Electrons.find({
```
Nitpick: I would skip a line after 68 for readability.

wbraynen commented on server_old.js:
```diff
@@ -0,0 +1,88 @@
+          { '$mod': [{'$subtract': ['$Angle', min]}, inc]}, 0]
+        },
+      ]}
+    },
+    function(err, docs) {
+      if (err) {
+        res.send(err)
+      } else {
+        res.json(docs)
+      }
+    }
+  )
```
Nitpick: formatting seems inconsistent with `fs.createReadStream` in the `uploadFunctions` function above.

Also, don't know how to make this better, but I don't understand what this does.  What's the lambda for?

wbraynen reviewed (commented):
Why is this PR specific to heroku?  Couldn't we run this code anywhere?  E.g. AWS, DigitalOcean, a university server, my laptop?

wbraynen commented:
Can we add or start thinking about how to add PR checks to this repo?  Unit tests or a linter or something; e.g. would this even run?  Seems a lot to rely purely on human review; could easily break what's in the main branch this way.

rexposadas commented:
> 

Sure thing.

rexposadas commented:
> Can we add or start thinking about how to add PR checks to this repo?

Sure thing.

wbraynen commented:
> The Procfile is what makes this specific to heroku. Admittedly, the branch is poorly named. It does a lot more than be deployable to Heroku.

I'm more worried about the PR title.  There are 13 commits in this PR, so guessing we'll want to squash them.  The easiest way to squash is to press Github's "Squash and merge" button (instead of using _interactive rebase_ locally and then force-pushing the changes).  If so, then the "Squash and merge" button will default the commit message to the PR title.

What should we change the PR title to?

wbraynen commented:
P.S. Trying to stay away from creating "merge commits" because a linear git history is easier on the eyes (easier to understand I think) and we don't need to back-merge anything (e.g. this is not a hotfix or a release branch where we are releasing from the release branch).

rexposadas commented:
> I'm more worried about the PR title.

I modified the title. 

I also removed the Profile for Heroku.  This keeps the PR focused on the CRUD api endpoints for 'projects'.

wbraynen reviewed (approved):
🎉 

Let's try to keep git history linear.  So anything but a _merge commit_.  So two options:
Option 1: "Merge and squash" this PR.  This will result in 1 commit on the `main` branch.
Option 2: Use _interactive rebase_ locally to squash commits to group them. This should result in more than 1 commit (but fewer than 14 commits).

wbraynen commented:
I haven't tested this PR.  Can someone else test it?  @katerynamelnykova can't install Docker on her laptop (b/c she has an incompatible version of Windows).  @DennisT98 Can you QA this?

DennisT98 commented:
Yes, I’ll test it later today.

DennisT98 commented:
Everything works fine for me, except for one thing. When running ```npm start```, I get an npm error for an invalid name: "A mock backend". This might be because my npm version (6.14.16) does not support spaces in this string, so I changed the name locally for testing purposes. No issues apart from that.
